mbed_lstools/platform_database.py

```diff
@@ -480,15 +489,101 @@ def remove(self, id, permanent=False, device_type='daplink', verbose_data=False)
+        Returns dictionary of all updates that occurred in the following format:
+            {
+                "new": {
+                    "<target id>": "<platform name>",
+                    ...
+                },
+                "updated": {
+                    "<target id>": {
+                        "old": "<previous platform name>",
+                        "new": "<new platform name>"
+                    }
+                }
+            }
+        """
         r = requests.get(url)
-        platform_data = r.json()
+        try:
+            platform_data = r.json()
+        except ValueError as e:
+            logger.debug("Failed to retrieve platform data from os.mbed.com.")
+            logger.debug(e)
+            logger.debug("Received the following response:")
+            logger.debug("Status code: %d", r.status_code)
+            logger.debug(r.text)
+            raise RemotePlatformDataException()
+
+        try:
+            """Expected JSON format is as follows:
+            [
+                {
+                    "productcode": "1234",
+                    "logicalboard": {
+                        "name": "boardname"
+                    }
+                },
+                ...
+            ]
+            """
+            web_platforms = {
+                v['productcode']: v['logicalboard']['name'].upper() for v in platform_data
+            }
+        except KeyError as e:
+            logger.debug("Platform data had unexpected format.")
+            logger.debug(e)
+            logger.debug("Received platform data:")
+            logger.debug(platform_data)
+            raise RemotePlatformDataException()
+
+        current_platforms = {
+            k: v
+            for (k, v) in self.items()
+        }
 
-        web_platforms = {
-            v['productcode']: v['logicalboard']['name'].upper() for v in platform_data
+        web_platform_keys = set(web_platforms.keys())
+        current_platform_keys = set(current_platforms.keys())
+
+        new_platform_ids = web_platform_keys - current_platform_keys
+        new_platforms = {
+            id: web_platforms[id]
+            for id in new_platform_ids
+        }
+
+        shared_platform_ids = web_platform_keys.intersection(current_platform_keys)
+
+        updated_platforms_history = {
+            id: {
+                "old": current_platforms[id],
+                "new": web_platforms[id]
+            }
+            for id in shared_platform_ids if web_platforms[id] != current_platforms[id]
         }
 
-        for id, platform_name in web_platforms.items():
+        updated_platforms = {
+            id: platform["new"]
+            for id, platform in updated_platforms_history.items()
+        }
+
+        all_updates = new_platforms.copy()
+        all_updates.update(updated_platforms)
+
+        for id, platform_name in all_updates.items():
             self.add(id, platform_name)
 
-        self._update_db()
+        if not self._update_db():
```

Comment:
why update_db does not raise exception ?

Reply:
I'm not seeing this part of the diff.

Reply:
It never has